src/addon/proxy.rs

```diff
@@ -0,0 +1,276 @@
+            );
+            return;
+        }
+
+        request
+            .headers_mut()
+            .append(HeaderName::from_static("via"), via_header);
+    }
+
+    /// Removes Hop-by-Hop headers which are only meaningful for a singles
+    /// transport-level connection and should not be stored by caches following
+    /// RFC2616.
+    ///
+    /// The following HTTP/1.1 headers are hop-by-hop headers:
+    ///
+    ///   - Connection
+    ///   - Keep-Alive
+    ///   - Proxy-Authenticate
+    ///   - Proxy-Authorization
+    ///   - TE
+    ///   - Trailers
+    ///   - Transfer-Encoding
+    ///   - Upgrade
+    ///
+    /// ## References
+    ///
+    /// http://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html
+    async fn remove_hbh_headers(&self, request: Request<Body>) {
+        use http::header::{
+            CONNECTION, PROXY_AUTHENTICATE, PROXY_AUTHORIZATION, TE, TRAILER, TRANSFER_ENCODING,
+            UPGRADE,
+        };
+
+        let mut request = request.lock().await;
+        let headers = request.headers_mut();
+
+        headers.remove(CONNECTION);
+        headers.remove(HeaderName::from_static("keep-alive"));
+        headers.remove(PROXY_AUTHENTICATE);
+        headers.remove(PROXY_AUTHORIZATION);
+        headers.remove(TE);
+        headers.remove(TRAILER);
+        headers.remove(TRANSFER_ENCODING);
+        headers.remove(UPGRADE);
+    }
+
+    /// Maps a _incoming_ HTTP request into a _outgoing_ HTTP request.
+    async fn map_incoming_request(&self, incoming: Request<Body>) -> hyper::Request<Body> {
+        let incoming = incoming.lock().await;
+        let upstream_uri = self.map_upstream_uri(incoming.uri());
+        let mut request = hyper::Request::builder()
+            .uri(upstream_uri)
+            .method(incoming.method())
+            .body(Body::empty())
+            .unwrap();
+        let headers = request.headers_mut();
+
+        *headers = incoming.headers().clone();
+
+        // TODO: Instead of append and removing it would be great to support
+        // some kind of `set` operation which adds if not present or replaces
+        // if present.
+        //
+        // Host must be the authority from the proxied URL
+        headers.remove(http::header::HOST);
+        headers.append(
+            http::header::HOST,
+            HeaderValue::from_str(self.upstream.authority().unwrap().as_str()).unwrap(),
+        );
+
+        // Specify Proxy as User Agent
+        headers.remove(USER_AGENT).unwrap();
+        headers.append(USER_AGENT, HeaderValue::from_static("Rust http-server/1.0"));
+
+        request
+    }
+
+    fn map_upstream_uri(&self, incoming_uri: &Uri) -> Uri {
+        let scheme = self.upstream.scheme_str().unwrap();
+        let authority = self.upstream.authority().unwrap().as_str();
+        let path_and_query = if let Some(paq) = incoming_uri.path_and_query() {
+            paq.as_str()
+        } else {
+            ""
+        };
```

Comment:
There must be a better way to achieve this. I will investigate on how to improve this piece.